--- a/configure
+++ b/configure
@@ -26747,7 +26747,7 @@ esac
 
 # Flags for building and linking against our internal copy of LuaJIT.
 
-LUAJIT_CPPFLAGS='-I$(top_srcdir)/lib/luajit/src'
+LUAJIT_CPPFLAGS='-I/usr/include/luajit-2.0'
 
 
 # We should be able to build http_load if epoll(2) is available.
--- a/configure.ac
+++ b/configure.ac
@@ -1401,7 +1401,7 @@ esac
 
 # Flags for building and linking against our internal copy of LuaJIT.
 AC_SUBST([LUAJIT_LDFLAGS])
-AC_SUBST([LUAJIT_CPPFLAGS], ['-I$(top_srcdir)/lib/luajit/src'])
+AC_SUBST([LUAJIT_CPPFLAGS], ['-I/usr/include/luajit-2.0'])
 
 # We should be able to build http_load if epoll(2) is available.
 AM_CONDITIONAL([BUILD_HTTP_LOAD], [test x"$ac_cv_func_epoll_ctl" = x"yes"])
