--- a/cmd/traffic_manager/Makefile.am
+++ b/cmd/traffic_manager/Makefile.am
@@ -29,7 +29,7 @@ AM_CPPFLAGS = \
   -I$(top_srcdir)/mgmt/cluster \
   -I$(top_srcdir)/mgmt/utils \
   -I$(top_srcdir)/lib \
-  -I$(top_srcdir)/lib/luajit/src
+  $(LUAJIT_CPPFLAGS)
 
 AM_LDFLAGS = \
   @EXTRA_CXX_LDFLAGS@ \
@@ -61,11 +61,10 @@ traffic_manager_SOURCES += \
   metrics.cc
 
 traffic_manager_LDADD += \
-  $(top_builddir)/lib/bindings/libbindings.la \
-  $(top_builddir)/lib/luajit/src/libluajit.a
+  $(top_builddir)/lib/bindings/libbindings.la
 
 traffic_manager_LDADD +=\
-  $(LIBUNWIND_LIBS) \
+  $(LIBUNWIND_LIBS) $(LUAJIT_LIBS) \
   @LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @LIBCAP@ @HWLOC_LIBS@ \
   -lm
 
--- a/cmd/traffic_manager/Makefile.in
+++ b/cmd/traffic_manager/Makefile.in
@@ -173,7 +173,6 @@ traffic_manager_DEPENDENCIES =  \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a \
 	$(top_builddir)/proxy/shared/libdiagsconfig.a \
 	$(top_builddir)/lib/bindings/libbindings.la \
-	$(top_builddir)/lib/luajit/src/libluajit.a \
 	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_2)
 AM_V_lt = $(am__v_lt_@AM_V@)
 am__v_lt_ = $(am__v_lt_@AM_DEFAULT_V@)
@@ -627,7 +626,7 @@ AM_CPPFLAGS = \
   -I$(top_srcdir)/mgmt/cluster \
   -I$(top_srcdir)/mgmt/utils \
   -I$(top_srcdir)/lib \
-  -I$(top_srcdir)/lib/luajit/src
+  $(LUAJIT_CPPFLAGS)
 
 AM_LDFLAGS = @EXTRA_CXX_LDFLAGS@ @LIBTOOL_LINK_FLAGS@ \
 	@OPENSSL_LDFLAGS@ $(LUAJIT_LDFLAGS)
@@ -641,9 +640,8 @@ traffic_manager_LDADD = $(top_builddir)/
 	$(top_builddir)/lib/ts/libtsutil.la \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a \
 	$(top_builddir)/proxy/shared/libdiagsconfig.a \
-	$(top_builddir)/lib/bindings/libbindings.la \
-	$(top_builddir)/lib/luajit/src/libluajit.a $(LIBUNWIND_LIBS) \
-	@LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @LIBCAP@ @HWLOC_LIBS@ -lm \
+	$(top_builddir)/lib/bindings/libbindings.la $(LIBUNWIND_LIBS) \
+	@LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @LIBCAP@ @HWLOC_LIBS@ $(LUAJIT_LIBS) -lm \
 	$(am__append_1)
 Clang_Tidy_Options = -fix -fix-errors
 
