--- a/proxy/Makefile.am
+++ b/proxy/Makefile.am
@@ -56,75 +56,6 @@ AM_LDFLAGS = \
 noinst_HEADERS = \
   Show.h
 
-# These are currently built separate, as part of building the lib/ tree, using
-# the normal LuaJIT build system. We are using the .o's directly, instead of the
-# luajit.a to avoid the linker from optimizing symbols away. We could maybe
-# switch to using the luajit.so, but that involves making sure it installs safely
-# and cleanly.
-LUAJIT = \
-  lib_aux.o \
-  lib_base.o \
-  lib_bit.o \
-  lib_debug.o \
-  lib_ffi.o \
-  lib_init.o \
-  lib_io.o \
-  lib_jit.o \
-  lib_math.o \
-  lib_os.o \
-  lib_package.o \
-  lib_string.o \
-  lib_table.o \
-  lj_alloc.o \
-  lj_api.o \
-  lj_asm.o \
-  lj_bc.o \
-  lj_bcread.o \
-  lj_bcwrite.o \
-  lj_carith.o \
-  lj_ccall.o \
-  lj_ccallback.o \
-  lj_cconv.o \
-  lj_cdata.o \
-  lj_char.o \
-  lj_clib.o \
-  lj_cparse.o \
-  lj_crecord.o \
-  lj_ctype.o \
-  lj_debug.o \
-  lj_dispatch.o \
-  lj_err.o \
-  lj_ffrecord.o \
-  lj_func.o \
-  lj_gc.o \
-  lj_gdbjit.o \
-  lj_ir.o \
-  lj_lex.o \
-  lj_lib.o \
-  lj_load.o \
-  lj_mcode.o \
-  lj_meta.o \
-  lj_obj.o \
-  lj_opt_dce.o \
-  lj_opt_fold.o \
-  lj_opt_loop.o \
-  lj_opt_mem.o \
-  lj_opt_narrow.o \
-  lj_opt_sink.o \
-  lj_opt_split.o \
-  lj_parse.o \
-  lj_record.o \
-  lj_snap.o \
-  lj_state.o \
-  lj_str.o \
-  lj_strscan.o \
-  lj_tab.o \
-  lj_trace.o \
-  lj_udata.o \
-  lj_vm.o \
-  lj_vmevent.o \
-  lj_vmmath.o
-
 EXTRA_DIST = InkAPITestTool.cc example_alarm_bin.sh example_prep.sh
 
 traffic_server_SOURCES = \
@@ -211,11 +142,11 @@ traffic_server_LDADD = \
   $(top_builddir)/iocore/aio/libinkaio.a \
   $(top_builddir)/lib/ts/libtsutil.la \
   $(top_builddir)/lib/bindings/libbindings.la \
-  $(top_builddir)/lib/luajit/src/libluajit.a \
   $(top_builddir)/iocore/net/libinknet.a \
   $(top_builddir)/iocore/eventsystem/libinkevent.a \
   $(top_builddir)/lib/records/librecords_p.a \
   $(top_builddir)/iocore/eventsystem/libinkevent.a \
+  $(LUAJIT_LIBS) \
   @HWLOC_LIBS@ \
   @LIBPCRE@ \
   @LIBTCL@ \
@@ -227,7 +158,6 @@ traffic_server_LDADD = \
   @OPENSSL_LIBS@ \
   -lm
 
-traffic_server_LDADD += $(LUAJIT:%=$(top_builddir)/lib/luajit/src/%)
 
 traffic_logcat_SOURCES = logcat.cc
 traffic_logcat_LDADD = \
@@ -239,9 +169,8 @@ traffic_logcat_LDADD = \
   $(top_builddir)/iocore/eventsystem/libinkevent.a \
   $(top_builddir)/lib/ts/libtsutil.la \
   $(top_builddir)/lib/bindings/libbindings.la \
-  $(top_builddir)/lib/luajit/src/libluajit.a \
   @LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @HWLOC_LIBS@\
-  @LIBPROFILER@ -lm
+  @LIBPROFILER@ $(LUAJIT_LIBS) -lm
 
 traffic_logstats_SOURCES = logstats.cc
 traffic_logstats_LDADD = \
@@ -253,9 +182,8 @@ traffic_logstats_LDADD = \
   $(top_builddir)/iocore/eventsystem/libinkevent.a \
   $(top_builddir)/lib/ts/libtsutil.la \
   $(top_builddir)/lib/bindings/libbindings.la \
-  $(top_builddir)/lib/luajit/src/libluajit.a \
   @LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @HWLOC_LIBS@ \
-  @LIBPROFILER@ -lm
+  @LIBPROFILER@ $(LUAJIT_LIBS) -lm
 
 versiondir = $(pkgsysconfdir)
 
--- a/proxy/Makefile.in
+++ b/proxy/Makefile.in
@@ -168,8 +168,7 @@ traffic_logcat_DEPENDENCIES = logging/li
 	$(top_builddir)/lib/records/librecords_p.a \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a \
 	$(top_builddir)/lib/ts/libtsutil.la \
-	$(top_builddir)/lib/bindings/libbindings.la \
-	$(top_builddir)/lib/luajit/src/libluajit.a
+	$(top_builddir)/lib/bindings/libbindings.la
 AM_V_lt = $(am__v_lt_@AM_V@)
 am__v_lt_ = $(am__v_lt_@AM_DEFAULT_V@)
 am__v_lt_0 = --silent
@@ -182,8 +181,7 @@ traffic_logstats_DEPENDENCIES = logging/
 	$(top_builddir)/lib/records/librecords_p.a \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a \
 	$(top_builddir)/lib/ts/libtsutil.la \
-	$(top_builddir)/lib/bindings/libbindings.la \
-	$(top_builddir)/lib/luajit/src/libluajit.a
+	$(top_builddir)/lib/bindings/libbindings.la
 am__traffic_server_SOURCES_DIST = AbstractBuffer.cc AbstractBuffer.h \
 	CacheControl.cc CacheControl.h ControlBase.cc ControlBase.h \
 	ControlMatcher.cc ControlMatcher.h CoreUtils.cc CoreUtils.h \
@@ -230,12 +228,10 @@ traffic_server_DEPENDENCIES = http/libht
 	$(top_builddir)/iocore/aio/libinkaio.a \
 	$(top_builddir)/lib/ts/libtsutil.la \
 	$(top_builddir)/lib/bindings/libbindings.la \
-	$(top_builddir)/lib/luajit/src/libluajit.a \
 	$(top_builddir)/iocore/net/libinknet.a \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a \
 	$(top_builddir)/lib/records/librecords_p.a \
-	$(top_builddir)/iocore/eventsystem/libinkevent.a \
-	$(LUAJIT:%=$(top_builddir)/lib/luajit/src/%)
+	$(top_builddir)/iocore/eventsystem/libinkevent.a
 traffic_server_LINK = $(LIBTOOL) $(AM_V_lt) --tag=CXX \
 	$(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=link $(CXXLD) \
 	$(AM_CXXFLAGS) $(CXXFLAGS) $(traffic_server_LDFLAGS) \
@@ -978,75 +974,6 @@ noinst_HEADERS = \
   Show.h
 
 
-# These are currently built separate, as part of building the lib/ tree, using
-# the normal LuaJIT build system. We are using the .o's directly, instead of the
-# luajit.a to avoid the linker from optimizing symbols away. We could maybe
-# switch to using the luajit.so, but that involves making sure it installs safely
-# and cleanly.
-LUAJIT = \
-  lib_aux.o \
-  lib_base.o \
-  lib_bit.o \
-  lib_debug.o \
-  lib_ffi.o \
-  lib_init.o \
-  lib_io.o \
-  lib_jit.o \
-  lib_math.o \
-  lib_os.o \
-  lib_package.o \
-  lib_string.o \
-  lib_table.o \
-  lj_alloc.o \
-  lj_api.o \
-  lj_asm.o \
-  lj_bc.o \
-  lj_bcread.o \
-  lj_bcwrite.o \
-  lj_carith.o \
-  lj_ccall.o \
-  lj_ccallback.o \
-  lj_cconv.o \
-  lj_cdata.o \
-  lj_char.o \
-  lj_clib.o \
-  lj_cparse.o \
-  lj_crecord.o \
-  lj_ctype.o \
-  lj_debug.o \
-  lj_dispatch.o \
-  lj_err.o \
-  lj_ffrecord.o \
-  lj_func.o \
-  lj_gc.o \
-  lj_gdbjit.o \
-  lj_ir.o \
-  lj_lex.o \
-  lj_lib.o \
-  lj_load.o \
-  lj_mcode.o \
-  lj_meta.o \
-  lj_obj.o \
-  lj_opt_dce.o \
-  lj_opt_fold.o \
-  lj_opt_loop.o \
-  lj_opt_mem.o \
-  lj_opt_narrow.o \
-  lj_opt_sink.o \
-  lj_opt_split.o \
-  lj_parse.o \
-  lj_record.o \
-  lj_snap.o \
-  lj_state.o \
-  lj_str.o \
-  lj_strscan.o \
-  lj_tab.o \
-  lj_trace.o \
-  lj_udata.o \
-  lj_vm.o \
-  lj_vmevent.o \
-  lj_vmmath.o
-
 EXTRA_DIST = InkAPITestTool.cc example_alarm_bin.sh example_prep.sh
 traffic_server_SOURCES = AbstractBuffer.cc AbstractBuffer.h \
 	CacheControl.cc CacheControl.h ControlBase.cc ControlBase.h \
@@ -1085,8 +1012,7 @@ traffic_server_LDADD = http/libhttp.a ht
 	$(top_builddir)/lib/records/librecords_p.a \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a @HWLOC_LIBS@ \
 	@LIBPCRE@ @LIBTCL@ @LIBCAP@ @LIBRESOLV@ @LIBZ@ @LIBLZMA@ \
-	@LIBPROFILER@ @OPENSSL_LIBS@ -lm \
-	$(LUAJIT:%=$(top_builddir)/lib/luajit/src/%)
+	@LIBPROFILER@ @OPENSSL_LIBS@ $(LUAJIT_LIBS) -lm
 traffic_logcat_SOURCES = logcat.cc
 traffic_logcat_LDADD = \
   logging/liblogging.a \
@@ -1099,7 +1025,7 @@ traffic_logcat_LDADD = \
   $(top_builddir)/lib/bindings/libbindings.la \
   $(top_builddir)/lib/luajit/src/libluajit.a \
   @LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @HWLOC_LIBS@\
-  @LIBPROFILER@ -lm
+  @LIBPROFILER@ $(LIBLUAJIT_LIBS) -lm
 
 traffic_logstats_SOURCES = logstats.cc
 traffic_logstats_LDADD = \
@@ -1113,7 +1039,7 @@ traffic_logstats_LDADD = \
   $(top_builddir)/lib/bindings/libbindings.la \
   $(top_builddir)/lib/luajit/src/libluajit.a \
   @LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @HWLOC_LIBS@ \
-  @LIBPROFILER@ -lm
+  @LIBPROFILER@ $(LUAJIT_LIBS) -lm
 
 versiondir = $(pkgsysconfdir)
 all: all-recursive
--- a/proxy/logging/Makefile.am
+++ b/proxy/logging/Makefile.am
@@ -21,7 +21,6 @@ include $(top_srcdir)/build/tidy.mk
 AM_CPPFLAGS = \
   $(iocore_include_dirs) \
   -I$(top_srcdir)/lib \
-  -I$(top_srcdir)/lib/luajit/src \
   -I$(top_srcdir)/lib/records \
   -I$(top_srcdir)/proxy/api/ts \
   -I$(top_srcdir)/proxy \
@@ -30,7 +29,8 @@ AM_CPPFLAGS = \
   -I$(top_srcdir)/proxy/hdrs \
   -I$(top_srcdir)/proxy/shared \
   -I$(top_srcdir)/mgmt \
-  -I$(top_srcdir)/mgmt/utils
+  -I$(top_srcdir)/mgmt/utils \
+  $(LUAJIT_CPPFLAGS)
 
 EXTRA_DIST = LogStandalone.cc
 
--- a/proxy/logging/Makefile.in
+++ b/proxy/logging/Makefile.in
@@ -841,7 +841,6 @@ CC_Clang_Tidy = $(CLANG_TIDY) $(Clang_Ti
 AM_CPPFLAGS = \
   $(iocore_include_dirs) \
   -I$(top_srcdir)/lib \
-  -I$(top_srcdir)/lib/luajit/src \
   -I$(top_srcdir)/lib/records \
   -I$(top_srcdir)/proxy/api/ts \
   -I$(top_srcdir)/proxy \
@@ -850,7 +849,8 @@ AM_CPPFLAGS = \
   -I$(top_srcdir)/proxy/hdrs \
   -I$(top_srcdir)/proxy/shared \
   -I$(top_srcdir)/mgmt \
-  -I$(top_srcdir)/mgmt/utils
+  -I$(top_srcdir)/mgmt/utils \
+  $(LUAJIT_CPPFLAGS)
 
 EXTRA_DIST = LogStandalone.cc
 noinst_LIBRARIES = liblogging.a liblogcollation.a
