--- a/cmd/traffic_manager/Makefile.am
+++ b/cmd/traffic_manager/Makefile.am
@@ -21,7 +21,7 @@ check_PROGRAMS = test_metrics
 TESTS = $(check_PROGRAMS)
 
 AM_CPPFLAGS = \
-  $(LUAJIT_CPPFLAGS) \
+  $(LUA_CPPFLAGS) \
   $(iocore_include_dirs) \
   -I$(abs_top_srcdir)/lib/records \
   -I$(abs_top_srcdir)/proxy/hdrs \
@@ -32,7 +32,7 @@ AM_CPPFLAGS = \
   -I$(abs_top_srcdir)/mgmt/cluster \
   -I$(abs_top_srcdir)/mgmt/utils \
   -I$(abs_top_srcdir)/lib \
-  -I$(abs_top_srcdir)/lib/luajit/src
+  $(LUA_CPPFLAGS)
 
 AM_LDFLAGS = \
   @EXTRA_CXX_LDFLAGS@ \
@@ -58,17 +58,16 @@ traffic_manager_LDADD = \
   $(top_builddir)/proxy/shared/libdiagsconfig.a
 
 AM_LDFLAGS += \
-  $(LUAJIT_LDFLAGS)
+  $(LUA_LDFLAGS)
 
 traffic_manager_SOURCES += \
   metrics.cc
 
 traffic_manager_LDADD += \
-  $(top_builddir)/lib/bindings/libbindings.la \
-  $(top_builddir)/lib/luajit/src/libluajit.a
+  $(top_builddir)/lib/bindings/libbindings.la
 
 traffic_manager_LDADD +=\
-  $(LIBUNWIND_LIBS) \
+  $(LIBUNWIND_LIBS) $(LUA_LIBS) \
   @LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @LIBCAP@ @HWLOC_LIBS@ \
   -lm
 
@@ -85,9 +84,9 @@ test_metrics_LDADD = \
   $(top_builddir)/mgmt/libmgmt_lm.la \
   $(top_builddir)/lib/records/librecords_lm.a \
   $(top_builddir)/lib/bindings/libbindings.la \
-  $(top_builddir)/lib/luajit/src/libluajit.a \
   $(top_builddir)/lib/ts/libtsutil.la \
   $(top_builddir)/iocore/eventsystem/libinkevent.a \
+  $(LUA_CPPFLAGS) \
   @LIBTCL@ @LIBPCRE@
 
 include $(top_srcdir)/build/tidy.mk
--- a/cmd/traffic_manager/Makefile.in
+++ b/cmd/traffic_manager/Makefile.in
@@ -163,7 +163,6 @@ test_metrics_OBJECTS = $(am_test_metrics
 test_metrics_DEPENDENCIES = $(top_builddir)/mgmt/libmgmt_lm.la \
 	$(top_builddir)/lib/records/librecords_lm.a \
 	$(top_builddir)/lib/bindings/libbindings.la \
-	$(top_builddir)/lib/luajit/src/libluajit.a \
 	$(top_builddir)/lib/ts/libtsutil.la \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a
 AM_V_lt = $(am__v_lt_@AM_V@)
@@ -187,7 +186,6 @@ traffic_manager_DEPENDENCIES =  \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a \
 	$(top_builddir)/proxy/shared/libdiagsconfig.a \
 	$(top_builddir)/lib/bindings/libbindings.la \
-	$(top_builddir)/lib/luajit/src/libluajit.a \
 	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_2)
 AM_V_P = $(am__v_P_@AM_V@)
 am__v_P_ = $(am__v_P_@AM_DEFAULT_V@)
@@ -567,9 +565,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
@@ -835,7 +833,7 @@ waith = @waith@
 zlibh = @zlibh@
 TESTS = $(check_PROGRAMS)
 AM_CPPFLAGS = \
-  $(LUAJIT_CPPFLAGS) \
+  $(LUA_CPPFLAGS) \
   $(iocore_include_dirs) \
   -I$(abs_top_srcdir)/lib/records \
   -I$(abs_top_srcdir)/proxy/hdrs \
@@ -846,10 +844,10 @@ AM_CPPFLAGS = \
   -I$(abs_top_srcdir)/mgmt/cluster \
   -I$(abs_top_srcdir)/mgmt/utils \
   -I$(abs_top_srcdir)/lib \
-  -I$(abs_top_srcdir)/lib/luajit/src
+  $(LUA_CPPFLAGS)
 
 AM_LDFLAGS = @EXTRA_CXX_LDFLAGS@ @LIBTOOL_LINK_FLAGS@ \
-	@OPENSSL_LDFLAGS@ $(LUAJIT_LDFLAGS)
+	@OPENSSL_LDFLAGS@ $(LUA_LDFLAGS)
 traffic_manager_SOURCES = AddConfigFilesHere.cc MgmtHandlers.cc \
 	MgmtHandlers.h WebOverview.cc WebOverview.h metrics.h \
 	traffic_manager.cc metrics.cc
@@ -860,18 +858,17 @@ traffic_manager_LDADD = $(top_builddir)/
 	$(top_builddir)/lib/ts/libtsutil.la \
 	$(top_builddir)/iocore/eventsystem/libinkevent.a \
 	$(top_builddir)/proxy/shared/libdiagsconfig.a \
-	$(top_builddir)/lib/bindings/libbindings.la \
-	$(top_builddir)/lib/luajit/src/libluajit.a $(LIBUNWIND_LIBS) \
-	@LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @LIBCAP@ @HWLOC_LIBS@ -lm \
+	$(top_builddir)/lib/bindings/libbindings.la $(LIBUNWIND_LIBS) \
+	@LIBRESOLV@ @LIBPCRE@ @LIBTCL@ @LIBCAP@ @HWLOC_LIBS@ $(LUA_LIBS) -lm \
 	$(am__append_1)
 test_metrics_SOURCES = test_metrics.cc metrics.cc WebOverview.cc
 test_metrics_LDADD = \
   $(top_builddir)/mgmt/libmgmt_lm.la \
   $(top_builddir)/lib/records/librecords_lm.a \
   $(top_builddir)/lib/bindings/libbindings.la \
-  $(top_builddir)/lib/luajit/src/libluajit.a \
   $(top_builddir)/lib/ts/libtsutil.la \
   $(top_builddir)/iocore/eventsystem/libinkevent.a \
+  $(LUA_CPPFLAGS)
   @LIBTCL@ @LIBPCRE@
 
 Clang_Tidy_Options = -fix -fix-errors
--- a/cmd/Makefile.in
+++ b/cmd/Makefile.in
@@ -507,9 +507,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
--- a/cmd/traffic_cop/Makefile.in
+++ b/cmd/traffic_cop/Makefile.in
@@ -315,9 +315,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
--- a/cmd/traffic_crashlog/Makefile.in
+++ b/cmd/traffic_crashlog/Makefile.in
@@ -335,9 +335,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
--- a/cmd/traffic_ctl/Makefile.in
+++ b/cmd/traffic_ctl/Makefile.in
@@ -318,9 +318,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
--- a/cmd/traffic_layout/Makefile.in
+++ b/cmd/traffic_layout/Makefile.in
@@ -315,9 +315,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
--- a/cmd/traffic_top/Makefile.in
+++ b/cmd/traffic_top/Makefile.in
@@ -315,9 +315,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
--- a/cmd/traffic_via/Makefile.in
+++ b/cmd/traffic_via/Makefile.in
@@ -518,9 +518,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
--- a/cmd/traffic_wccp/Makefile.in
+++ b/cmd/traffic_wccp/Makefile.in
@@ -315,9 +315,9 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
