Index: trafficserver/lib/Makefile.in
===================================================================
--- trafficserver.orig/lib/Makefile.in	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/Makefile.in	2018-01-21 13:07:23.657375608 +0100
@@ -305,9 +305,9 @@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
@@ -912,29 +912,6 @@
 .PRECIOUS: Makefile
 
 
-# Some special hacks around building the luajit.
-#
-#  - Copy the luajit source tree if we're doing out-of-source builds
-#  - Upon clean, remove the luajit tree from the out-of-source build tree
-#
-# Note: The LUAJIT_LDFLAGS is a bit of a hack, since LuaJIT on OmniOS needs
-#       the -m64 flag. See configure.ac.
-#
-all-local:
-	@echo "Making all in luajit"
-	test -d "$(top_builddir)/$(subdir)/luajit/src" || cp -rf "$(srcdir)/luajit" "$(top_builddir)/$(subdir)/"
-	cd luajit && $(MAKE) $(AM_MAKEFLAGS) \
-		BUILDMODE="static" \
-		PREFIX="$(prefix)" \
-		CC="$(CC)" \
-		CFLAGS="$(LUAJIT_CFLAGS)" \
-		XCFLAGS="" \
-		LDFLAGS="$(LUAJIT_LDFLAGS)" \
-		Q=$(luajit__v_$(V))
-
-clean-local:
-	test "$(top_srcdir)" != "$(top_builddir)" || (cd "$(top_builddir)/$(subdir)/luajit" && $(MAKE) clean)
-	test "$(top_srcdir)" = "$(top_builddir)" || rm -rf "$(top_builddir)/$(subdir)/luajit"
 
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
Index: trafficserver/lib/bindings/Makefile.am
===================================================================
--- trafficserver.orig/lib/bindings/Makefile.am	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/bindings/Makefile.am	2018-01-21 13:07:23.657375608 +0100
@@ -19,9 +19,10 @@
 include $(top_srcdir)/build/tidy.mk
 
 AM_CPPFLAGS += \
-  $(LUAJIT_CPPFLAGS) \
+  $(LUA_CPPFLAGS) \
   -I$(abs_top_srcdir)/lib \
-  -I$(abs_top_srcdir)/lib/records
+  -I$(abs_top_srcdir)/lib/records \
+  $(LUA_CPPFLAGS)
 
 noinst_LTLIBRARIES = libbindings.la
 
Index: trafficserver/lib/bindings/Makefile.in
===================================================================
--- trafficserver.orig/lib/bindings/Makefile.in	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/bindings/Makefile.in	2018-01-21 13:07:23.657375608 +0100
@@ -243,7 +243,7 @@
 ACLOCAL = @ACLOCAL@
 AMTAR = @AMTAR@
 AM_CFLAGS = @AM_CFLAGS@
-AM_CPPFLAGS = @AM_CPPFLAGS@ $(LUAJIT_CPPFLAGS) -I$(abs_top_srcdir)/lib \
+AM_CPPFLAGS = @AM_CPPFLAGS@ $(LUA_CPPFLAGS) -I$(abs_top_srcdir)/lib \
 	-I$(abs_top_srcdir)/lib/records
 AM_CXXFLAGS = @AM_CXXFLAGS@
 AM_DEFAULT_VERBOSITY = @AM_DEFAULT_VERBOSITY@
@@ -327,9 +327,9 @@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
Index: trafficserver/lib/perl/Makefile.in
===================================================================
--- trafficserver.orig/lib/perl/Makefile.in	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/perl/Makefile.in	2018-01-21 13:07:23.657375608 +0100
@@ -242,9 +242,9 @@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
Index: trafficserver/lib/records/Makefile.in
===================================================================
--- trafficserver.orig/lib/records/Makefile.in	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/records/Makefile.in	2018-01-21 13:07:23.661375620 +0100
@@ -350,9 +350,9 @@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
Index: trafficserver/lib/ts/Makefile.in
===================================================================
--- trafficserver.orig/lib/ts/Makefile.in	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/ts/Makefile.in	2018-01-21 13:07:23.661375620 +0100
@@ -622,9 +622,9 @@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
Index: trafficserver/lib/tsconfig/Makefile.in
===================================================================
--- trafficserver.orig/lib/tsconfig/Makefile.in	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/tsconfig/Makefile.in	2018-01-21 13:07:23.661375620 +0100
@@ -399,9 +399,9 @@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
Index: trafficserver/lib/wccp/Makefile.in
===================================================================
--- trafficserver.orig/lib/wccp/Makefile.in	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/wccp/Makefile.in	2018-01-21 13:07:23.661375620 +0100
@@ -335,9 +335,9 @@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 LT_SYS_LIBRARY_PATH = @LT_SYS_LIBRARY_PATH@
-LUAJIT_CFLAGS = @LUAJIT_CFLAGS@
-LUAJIT_CPPFLAGS = @LUAJIT_CPPFLAGS@
-LUAJIT_LDFLAGS = @LUAJIT_LDFLAGS@
+LUA_CFLAGS = @LUA_CFLAGS@
+LUA_CPPFLAGS = @LUA_CPPFLAGS@
+LUA_LDFLAGS = @LUA_LDFLAGS@
 MAINT = @MAINT@
 MAKEINFO = @MAKEINFO@
 MANIFEST_TOOL = @MANIFEST_TOOL@
Index: trafficserver/lib/Makefile.am
===================================================================
--- trafficserver.orig/lib/Makefile.am	2018-01-21 13:07:23.665375633 +0100
+++ trafficserver/lib/Makefile.am	2018-01-21 13:07:23.661375620 +0100
@@ -25,32 +25,3 @@
 if BUILD_WCCP
 SUBDIRS += wccp
 endif
-
-# Support verbose LuaJIT builds with "make V=1".
-luajit__v_ = $(luajit__v_$(AM_DEFAULT_VERBOSITY))
-luajit__v_0 = @
-luajit__v_1 =
-
-# Some special hacks around building the luajit.
-#
-#  - Copy the luajit source tree if we're doing out-of-source builds
-#  - Upon clean, remove the luajit tree from the out-of-source build tree
-#
-# Note: The LUAJIT_LDFLAGS is a bit of a hack, since LuaJIT on OmniOS needs
-#       the -m64 flag. See configure.ac.
-#
-all-local:
-	@echo "Making all in luajit"
-	test -d "$(top_builddir)/$(subdir)/luajit/src" || cp -rf "$(srcdir)/luajit" "$(top_builddir)/$(subdir)/"
-	cd luajit && $(MAKE) $(AM_MAKEFLAGS) \
-		BUILDMODE="static" \
-		PREFIX="$(prefix)" \
-		CC="$(CC)" \
-		CFLAGS="$(LUAJIT_CFLAGS)" \
-		XCFLAGS="" \
-		LDFLAGS="$(LUAJIT_LDFLAGS)" \
-		Q=$(luajit__v_$(V))
-
-clean-local:
-	test "$(top_srcdir)" != "$(top_builddir)" || (cd "$(top_builddir)/$(subdir)/luajit" && $(MAKE) clean)
-	test "$(top_srcdir)" = "$(top_builddir)" || rm -rf "$(top_builddir)/$(subdir)/luajit"
