--- a/proxy/logging/Log.cc
+++ b/proxy/logging/Log.cc
@@ -1300,7 +1300,9 @@ Log::flush_thread_main(void * /* args AT
       // make sure we're open & ready to write
       logfile->check_fd();
       if (!logfile->is_open()) {
-        Warning("File:%s was closed, have dropped (%d) bytes.", logfile->get_name(), total_bytes);
+        if (logfile->m_file_format != LOG_FILE_PIPE) {
+          Warning("File:%s was closed, have dropped (%d) bytes.", logfile->get_name(), total_bytes);
+        }
 
         RecIncrRawStat(log_rsb, mutex->thread_holding, log_stat_bytes_lost_before_written_to_disk_stat, total_bytes);
         delete fdata;
@@ -1326,8 +1328,10 @@ Log::flush_thread_main(void * /* args AT
         len = ::write(logfilefd, &buf[bytes_written], total_bytes - bytes_written);
 
         if (len < 0) {
-          Error("Failed to write log to %s: [tried %d, wrote %d, %s]", logfile->get_name(), total_bytes - bytes_written,
-                bytes_written, strerror(errno));
+          if (logfile->m_file_format != LOG_FILE_PIPE) {
+            Error("Failed to write log to %s: [tried %d, wrote %d, %s]", logfile->get_name(), total_bytes - bytes_written,
+                  bytes_written, strerror(errno));
+          }
 
           RecIncrRawStat(log_rsb, mutex->thread_holding, log_stat_bytes_lost_before_written_to_disk_stat,
                          total_bytes - bytes_written);
