--- a/include/ts/ts.h
+++ b/include/ts/ts.h
@@ -2242,6 +2242,31 @@ tsapi int TSHttpTxnServerRespHdrBytesGet
 tsapi int64_t TSHttpTxnServerRespBodyBytesGet(TSHttpTxn txnp);
 tsapi int TSHttpTxnClientRespHdrBytesGet(TSHttpTxn txnp);
 tsapi int64_t TSHttpTxnClientRespBodyBytesGet(TSHttpTxn txnp);
+tsapi int TSVConnIsSslReused(TSVConn sslp);
+
+/**
+   Return the current (if set) SSL Cipher. This is still owned by the
+   core, and must not be free'd.
+   sslp The connection pointer
+   @return the SSL Cipher
+*/
+tsapi const char *TSVConnSslCipherGet(TSVConn sslp);
+
+/**
+   Return the current (if set) SSL Protocol. This is still owned by the
+   core, and must not be free'd.
+   sslp The connection pointer
+   @return the SSL Protocol
+*/
+tsapi const char *TSVConnSslProtocolGet(TSVConn sslp);
+
+/**
+   Return the current (if set) SSL Curve. This is still owned by the
+   core, and must not be free'd.
+   sslp The connection pointer
+   @return the SSL Curve
+*/
+tsapi const char *TSVConnSslCurveGet(TSVConn sslp);
 
 /* NetVC timeout APIs. */
 tsapi void TSVConnInactivityTimeoutSet(TSVConn connp, TSHRTime timeout);
--- a/plugins/lua/ts_lua_client_request.c
+++ b/plugins/lua/ts_lua_client_request.c
@@ -66,6 +66,15 @@ static int ts_lua_client_request_client_
 static int ts_lua_client_request_client_addr_get_addr(lua_State *L);
 static int ts_lua_client_request_client_addr_get_incoming_port(lua_State *L);
 
+static void ts_lua_inject_client_request_ssl_reused_api(lua_State *L);
+static int ts_lua_client_request_get_ssl_reused(lua_State *L);
+static void ts_lua_inject_client_request_ssl_cipher_api(lua_State *L);
+static int ts_lua_client_request_get_ssl_cipher(lua_State *L);
+static void ts_lua_inject_client_request_ssl_protocol_api(lua_State *L);
+static int ts_lua_client_request_get_ssl_protocol(lua_State *L);
+static void ts_lua_inject_client_request_ssl_curve_api(lua_State *L);
+static int ts_lua_client_request_get_ssl_curve(lua_State *L);
+
 void
 ts_lua_inject_client_request_api(lua_State *L)
 {
@@ -82,6 +91,10 @@ ts_lua_inject_client_request_api(lua_Sta
   ts_lua_inject_client_request_version_api(L);
   ts_lua_inject_client_request_body_size_api(L);
   ts_lua_inject_client_request_header_size_api(L);
+  ts_lua_inject_client_request_ssl_reused_api(L);
+  ts_lua_inject_client_request_ssl_cipher_api(L);
+  ts_lua_inject_client_request_ssl_protocol_api(L);
+  ts_lua_inject_client_request_ssl_curve_api(L);
 
   lua_setfield(L, -2, "client_request");
 }
@@ -923,3 +936,115 @@ ts_lua_client_request_get_header_size(lu
 
   return 1;
 }
+
+
+static void
+ts_lua_inject_client_request_ssl_reused_api(lua_State *L)
+{
+  lua_pushcfunction(L, ts_lua_client_request_get_ssl_reused);
+  lua_setfield(L, -2, "get_ssl_reused");
+}
+
+static int
+ts_lua_client_request_get_ssl_reused(lua_State *L)
+{
+  int ssl_reused = 0;
+  ts_lua_http_ctx *http_ctx;
+  TSHttpSsn ssnp;
+  TSVConn client_conn;
+
+  GET_HTTP_CONTEXT(http_ctx, L);
+  ssnp        = TSHttpTxnSsnGet(http_ctx->txnp);
+  client_conn = TSHttpSsnClientVConnGet(ssnp);
+
+  if (TSVConnIsSsl(client_conn)) {
+    ssl_reused = TSVConnIsSslReused(client_conn);
+  }
+  lua_pushnumber(L, ssl_reused);
+
+  return 1;
+}
+
+static void
+ts_lua_inject_client_request_ssl_cipher_api(lua_State *L)
+{
+  lua_pushcfunction(L, ts_lua_client_request_get_ssl_cipher);
+  lua_setfield(L, -2, "get_ssl_cipher");
+}
+
+static int
+ts_lua_client_request_get_ssl_cipher(lua_State *L)
+{
+  const char *ssl_cipher = "-";
+  ts_lua_http_ctx *http_ctx;
+  TSHttpSsn ssnp;
+  TSVConn client_conn;
+
+  GET_HTTP_CONTEXT(http_ctx, L);
+  ssnp        = TSHttpTxnSsnGet(http_ctx->txnp);
+  client_conn = TSHttpSsnClientVConnGet(ssnp);
+
+  if (TSVConnIsSsl(client_conn)) {
+    ssl_cipher = TSVConnSslCipherGet(client_conn);
+  }
+
+  lua_pushstring(L, ssl_cipher);
+
+  return 1;
+}
+
+static void
+ts_lua_inject_client_request_ssl_protocol_api(lua_State *L)
+{
+  lua_pushcfunction(L, ts_lua_client_request_get_ssl_protocol);
+  lua_setfield(L, -2, "get_ssl_protocol");
+}
+
+static int
+ts_lua_client_request_get_ssl_protocol(lua_State *L)
+{
+  const char *ssl_protocol = "-";
+  ts_lua_http_ctx *http_ctx;
+  TSHttpSsn ssnp;
+  TSVConn client_conn;
+
+  GET_HTTP_CONTEXT(http_ctx, L);
+  ssnp        = TSHttpTxnSsnGet(http_ctx->txnp);
+  client_conn = TSHttpSsnClientVConnGet(ssnp);
+
+  if (TSVConnIsSsl(client_conn)) {
+    ssl_protocol = TSVConnSslProtocolGet(client_conn);
+  }
+
+  lua_pushstring(L, ssl_protocol);
+
+  return 1;
+}
+
+static void
+ts_lua_inject_client_request_ssl_curve_api(lua_State *L)
+{
+  lua_pushcfunction(L, ts_lua_client_request_get_ssl_curve);
+  lua_setfield(L, -2, "get_ssl_curve");
+}
+
+static int
+ts_lua_client_request_get_ssl_curve(lua_State *L)
+{
+  const char *ssl_curve = "-";
+  ts_lua_http_ctx *http_ctx;
+  TSHttpSsn ssnp;
+  TSVConn client_conn;
+
+  GET_HTTP_CONTEXT(http_ctx, L);
+  ssnp        = TSHttpTxnSsnGet(http_ctx->txnp);
+  client_conn = TSHttpSsnClientVConnGet(ssnp);
+
+  if (TSVConnIsSsl(client_conn)) {
+    ssl_curve = TSVConnSslCurveGet(client_conn);
+  }
+
+  lua_pushstring(L, ssl_curve);
+
+  return 1;
+}
--- a/src/traffic_server/InkAPI.cc
+++ b/src/traffic_server/InkAPI.cc
@@ -6192,6 +6192,42 @@ TSHttpTxnClientRespBodyBytesGet(TSHttpTx
 }
 
 int
+TSVConnIsSslReused(TSVConn sslp)
+{
+  NetVConnection *vc        = reinterpret_cast<NetVConnection *>(sslp);
+  SSLNetVConnection *ssl_vc = dynamic_cast<SSLNetVConnection *>(vc);
+
+  return ssl_vc ? ssl_vc->getSSLSessionCacheHit() : 0;
+}
+
+const char *
+TSVConnSslCipherGet(TSVConn sslp)
+{
+  NetVConnection *vc        = reinterpret_cast<NetVConnection *>(sslp);
+  SSLNetVConnection *ssl_vc = dynamic_cast<SSLNetVConnection *>(vc);
+
+  return ssl_vc ? ssl_vc->getSSLCipherSuite() : nullptr;
+}
+
+const char *
+TSVConnSslProtocolGet(TSVConn sslp)
+{
+  NetVConnection *vc        = reinterpret_cast<NetVConnection *>(sslp);
+  SSLNetVConnection *ssl_vc = dynamic_cast<SSLNetVConnection *>(vc);
+
+  return ssl_vc ? ssl_vc->getSSLProtocol() : nullptr;
+}
+
+const char *
+TSVConnSslCurveGet(TSVConn sslp)
+{
+  NetVConnection *vc        = reinterpret_cast<NetVConnection *>(sslp);
+  SSLNetVConnection *ssl_vc = dynamic_cast<SSLNetVConnection *>(vc);
+
+  return ssl_vc ? ssl_vc->getSSLCurve() : nullptr;
+}
+
+int
 TSHttpTxnPushedRespHdrBytesGet(TSHttpTxn txnp)
 {
   sdk_assert(sdk_sanity_check_txn(txnp) == TS_SUCCESS);
