Origin: upstream https://github.com/apache/trafficserver/pull/5888
Description: Allow disabling HTTP/2 priority frames limit
--- a/proxy/http2/Http2ConnectionState.cc
+++ b/proxy/http2/Http2ConnectionState.cc
@@ -406,10 +406,15 @@ rcv_priority_frame(Http2ConnectionState 
                       "PRIORITY frame depends on itself");
   }
 
+  if (!Http2::stream_priority_enabled) {
+    return Http2Error(Http2ErrorClass::HTTP2_ERROR_CLASS_NONE);
+  }
+
   // Update PRIORITY frame count per minute
   cstate.increment_received_priority_frame_count();
   // Close this conection if its priority frame count received exceeds a limit
-  if (cstate.get_received_priority_frame_count() > Http2::max_priority_frames_per_minute) {
+  if (Http2::max_priority_frames_per_minute != 0 &&
+      cstate.get_received_priority_frame_count() > Http2::max_priority_frames_per_minute) {
     Http2StreamDebug(cstate.ua_session, stream_id,
                      "Observed too frequent priority changes: %u priority changes within a last minute",
                      cstate.get_received_priority_frame_count());
@@ -417,10 +422,6 @@ rcv_priority_frame(Http2ConnectionState 
                       "recv priority too frequent priority changes");
   }
 
-  if (!Http2::stream_priority_enabled) {
-    return Http2Error(Http2ErrorClass::HTTP2_ERROR_CLASS_NONE);
-  }
-
   Http2StreamDebug(cstate.ua_session, stream_id, "PRIORITY - dep: %d, weight: %d, excl: %d, tree size: %d",
                    priority.stream_dependency, priority.weight, priority.exclusive_flag, cstate.dependency_tree->size());
 
