Origin: upstream https://github.com/apache/trafficserver/pull/5153
Description: Separate P_SSLUtil and P_SSLClientUtils includes
--- a/iocore/net/P_SSLUtils.h
+++ b/iocore/net/P_SSLUtils.h
@@ -23,7 +23,8 @@
 
 #include "tscore/ink_config.h"
 #include "tscore/Diags.h"
-#include "P_SSLClientUtils.h"
+#include "records/I_RecCore.h"
+#include "tscore/Map.h"
 
 #define OPENSSL_THREAD_DEFINES
 
--- a/iocore/net/SSLConfig.cc
+++ b/iocore/net/SSLConfig.cc
@@ -37,6 +37,7 @@
 #include "P_Net.h"
 #include "P_SSLConfig.h"
 #include "P_SSLUtils.h"
+#include "P_SSLClientUtils.h"
 #include "P_SSLCertLookup.h"
 #include "SSLSessionCache.h"
 #include <records/I_RecHttp.h>
--- a/src/traffic_server/traffic_server.cc
+++ b/src/traffic_server/traffic_server.cc
@@ -96,6 +96,7 @@ extern "C" int plock(int);
 #include "HTTP2.h"
 #include "tscore/ink_config.h"
 #include "P_SSLSNI.h"
+#include "P_SSLClientUtils.h"
 
 #include "tscore/ink_cap.h"
 
