--- a/plugins/experimental/Makefile.am
+++ b/plugins/experimental/Makefile.am
@@ -16,14 +16,18 @@
 
 if BUILD_EXPERIMENTAL_PLUGINS
 SUBDIRS = \
- lua \
- esi \
- rfc5861 \
- tcp_info \
+ buffer_upload \
+ cacheurl \
+ channel_stats \
  custom_redirect \
+ esi \
+ geoip_acl \
+ gzip \
  header_rewrite \
+ lua \
  metalink \
- gzip \
+ mysql_remap \
+ rfc5861 \
  spdy \
- channel_stats
+ tcp_info
 endif
--- a/configure.ac
+++ b/configure.ac
@@ -1605,16 +1605,20 @@
 AC_CONFIG_FILES([plugins/stats_over_http/Makefile])
 # experimental plugins
 AC_CONFIG_FILES([plugins/experimental/Makefile])
-AC_CONFIG_FILES([plugins/experimental/esi/Makefile])
-AC_CONFIG_FILES([plugins/experimental/lua/Makefile])
-AC_CONFIG_FILES([plugins/experimental/rfc5861/Makefile])
-AC_CONFIG_FILES([plugins/experimental/tcp_info/Makefile])
+AC_CONFIG_FILES([plugins/experimental/buffer_upload/Makefile])
+AC_CONFIG_FILES([plugins/experimental/cacheurl/Makefile])
+AC_CONFIG_FILES([plugins/experimental/channel_stats/Makefile])
 AC_CONFIG_FILES([plugins/experimental/custom_redirect/Makefile])
+AC_CONFIG_FILES([plugins/experimental/esi/Makefile])
+AC_CONFIG_FILES([plugins/experimental/geoip_acl/Makefile])
+AC_CONFIG_FILES([plugins/experimental/gzip/Makefile])
 AC_CONFIG_FILES([plugins/experimental/header_rewrite/Makefile])
+AC_CONFIG_FILES([plugins/experimental/lua/Makefile])
 AC_CONFIG_FILES([plugins/experimental/metalink/Makefile])
-AC_CONFIG_FILES([plugins/experimental/gzip/Makefile])
+AC_CONFIG_FILES([plugins/experimental/mysql_remap/Makefile])
+AC_CONFIG_FILES([plugins/experimental/rfc5861/Makefile])
 AC_CONFIG_FILES([plugins/experimental/spdy/Makefile])
-AC_CONFIG_FILES([plugins/experimental/channel_stats/Makefile])
+AC_CONFIG_FILES([plugins/experimental/tcp_info/Makefile])
 # various tools
 AC_CONFIG_FILES([tools/Makefile])
 # example plugins
--- /dev/null
+++ b/plugins/experimental/buffer_upload/Makefile.am
@@ -0,0 +1,22 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+AM_CPPFLAGS = -I$(top_builddir)/proxy/api -I$(top_srcdir)/proxy/api
+pkglibdir = ${pkglibexecdir}
+
+pkglib_LTLIBRARIES = buffer_upload.la
+buffer_upload_la_SOURCES = buffer_upload.cc
+buffer_upload_la_LDFLAGS = -avoid-version -module -shared
--- /dev/null
+++ b/plugins/experimental/cacheurl/Makefile.am
@@ -0,0 +1,22 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+AM_CPPFLAGS = -I$(top_builddir)/proxy/api -I$(top_srcdir)/proxy/api
+pkglibdir = ${pkglibexecdir}
+
+pkglib_LTLIBRARIES = cacheurl.la
+cacheurl_la_SOURCES = cacheurl.c
+cacheurl_la_LDFLAGS = -avoid-version -module -shared
--- /dev/null
+++ b/plugins/experimental/geoip_acl/Makefile.am
@@ -0,0 +1,22 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+AM_CPPFLAGS = -I$(top_builddir)/proxy/api -I$(top_srcdir)/proxy/api
+pkglibdir = ${pkglibexecdir}
+
+pkglib_LTLIBRARIES = geoip_acl.la
+geoip_acl_la_SOURCES = geoip_acl.cc acl.cc
+geoip_acl_la_LDFLAGS = -avoid-version -module -shared
--- /dev/null
+++ b/plugins/experimental/mysql_remap/Makefile.am
@@ -0,0 +1,37 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+pkglibdir = ${pkglibexecdir}
+
+
+AM_CXXFLAGS = \
+  -I$(srcdir)/lib \
+  -I$(top_builddir)/proxy/api \
+  -I$(top_srcdir)/proxy/api \
+  -Wno-deprecated
+
+pkglib_LTLIBRARIES = mysql_remap.la
+lib_LTLIBRARIES = libmysql_remap.la
+
+libmysql_remap_la_SOURCES = \
+	lib/dictionary.c \
+	lib/iniparser.c
+
+mysql_remap_la_SOURCES =  \
+	mysql_remap.cc
+
+mysql_remap_la_LIBADD = libmysql_remap.la
+mysql_remap_la_LDFLAGS = -module -avoid-version -shared
