Origin: <vendor|upstream|other>, <url of original patch>
Bug: <url in upstream bugtracker>
Bug-Debian: http://bugs.debian.org/<bugnumber>
Bug-Ubuntu: https://launchpad.net/bugs/<bugnumber>
Forwarded: <no|not-needed|url proving that it has been forwarded>
Reviewed-By: <name and email of someone who approved the patch>
Last-Update: <YYYY-MM-DD>

--- a/configure.ac
+++ b/configure.ac
@@ -543,7 +543,7 @@
     TS_ADDTO(CPPFLAGS, [-I/opt/local/include])
     TS_ADDTO(LDFLAGS, [-L/opt/local/lib])
     ;;
-  freebsd*)
+  freebsd*|kfreebsd*)
     common_opt="-pipe -Wall -Werror"
     debug_opt="-ggdb3 $common_opt"
     release_opt="-g $common_opt -O3 -feliminate-unused-debug-symbols -fno-strict-aliasing"
@@ -688,6 +688,11 @@
     EXTRA_CXX_LDFLAGS="-rdynamic"
     host_os_def=freebsd
     ;;
+  kfreebsd*)
+    EXTRA_CXX_LDFLAGS="-rdynamic"
+    host_os_def=freebsd
+    TS_ADDTO(CPPFLAGS, [-Dkfreebsd])
+    ;;
   solaris*)
     host_os_def=solaris
     ;;
@@ -698,6 +703,8 @@
 esac
 TS_ADDTO(CPPFLAGS, [-D$host_os_def])
 
+AC_MSG_NOTICE([Build for host OS: $host_os, arch: $host_cpu, optimization: $host_os_def])
+
 #
 # _Here_ is where we go ahead and add the _optimizations_ to already
 #  existing CFLAGS/CXXFLAGS if some special values had been set.
@@ -776,7 +783,7 @@
 AC_SUBST([LIBTCL],[$TCL_LIB_FLAG])
 
 case $host_os in
-  freebsd*)
+  freebsd*|kfreebsd*)
     TS_ADDTO(CPPFLAGS, [-I/usr/local/include])
     ;;
   solaris*)
--- a/proxy/CoreUtils.cc
+++ b/proxy/CoreUtils.cc
@@ -979,7 +979,7 @@
 
     // This is not 64-bit correct. /leif
     printf("----------- EThread @ 0x%p ----------\n", eth_test);
-#if defined(freebsd) || defined(darwin)
+#if defined(darwin)
     printf("   thread_id: %p\n", loaded_eth->tid);
 #else
     printf("   thread_id: %i\n", (int) loaded_eth->tid);
--- a/cop/TrafficCop.cc
+++ b/cop/TrafficCop.cc
@@ -728,7 +728,7 @@
     if (err < 0) {
       break;
     }
-#if !defined(linux) && !defined(freebsd) && !defined(darwin)
+#if defined(solaris) || defined(kfreebsd) || defined(unknown)
     err = semctl(err, 1, IPC_RMID);
 #else
     union semun dummy_semun;
@@ -1975,7 +1975,7 @@
   signal(SIGTTIN, SIG_IGN);
 
   setsid();                     // Important, thanks Vlad. :)
-#if defined(freebsd)
+#if defined(freebsd) && !defined(kfreebsd)
   setpgrp(0,0);
 #else
   setpgrp();
--- a/lib/ts/ink_string.cc
+++ b/lib/ts/ink_string.cc
@@ -500,7 +500,7 @@
 
   inbytesleft = inlen;
   outbytesleft = *outlen;
-#if defined(freebsd) || defined(solaris)
+#if !defined(kfreebsd) && (defined(freebsd) || defined(solaris))
   if (iconv(ic, &in, &inbytesleft, &out, &outbytesleft) == (size_t) - 1)
 #else
   if (iconv(ic, (char **) &in, &inbytesleft, &out, &outbytesleft) == (size_t) - 1)
